src/OrleansRuntime/LogViews/GrainStateWithMetaData.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime.LogViews
+{
+    /// <summary>
+    /// Used by StorageProviderLogViewAdaptor, which allows a plain storage provider to
+    /// be wrapped by a log view provider, by adding meta data to the grain state.
+    /// </summary>
+    /// <typeparam name="TView">The view state</typeparam>
+    [Serializable]
+    public class GrainStateWithMetaDataAndETag<TView> : IGrainState where TView: class, new()
+    {
+        public GrainStateWithMetaData<TView> StateAndMetaData { get; set; }
+       
+        public string ETag { get; set; }
+
+        object IGrainState.State
+        {
+            get
+            {
+                return StateAndMetaData;
+            }
+            set
+            {
+                StateAndMetaData = (GrainStateWithMetaData<TView>) value;
+            }
+        }
+
+        public GrainStateWithMetaDataAndETag(TView initialview)
+        {
+            StateAndMetaData = new GrainStateWithMetaData<TView>(initialview);
+        }
+        public GrainStateWithMetaDataAndETag()
+        {
+            StateAndMetaData = new GrainStateWithMetaData<TView>();
+        }
+
+        public override string ToString()
+        {
+            return string.Format("v{0} Flags={1} ETag={2} Data={3}", StateAndMetaData.GlobalVersion, StateAndMetaData.WriteVector, ETag, StateAndMetaData.State);
+        }
+    }
+
+
+    [Serializable]
+    public class GrainStateWithMetaData<TView> where TView : class, new()
+    {
+        /// <summary>
+        /// The stored view of the log
+        /// </summary>
+        public TView State { get; set; }
+
+        /// <summary>
+        /// The length of the log
+        /// </summary>
+        public int GlobalVersion { get; set; }
+
+
+        /// <summary>
+        /// Used to avoid duplicate appends.
+        /// A string representing a bit vector, with one bit per replica. 
+        /// Bits are toggled when writing, so that the retry logic can avoid appending an entry twice
+        /// when retrying a failed append. 
+        /// </summary>
+        public string WriteVector { get; set; }
```

Comment:
Understood.  Thanks for the clarification.
If I understand you're goals, you want to avoid the cost in cpu and space of serializing a dictionary, and you want the contents to be in a human readable format when written to storage.  Is this correct?
If so, I suggest two properties.  One for programmatic use, one for storage.  Something like:

    [NonSerialized]
    public HashSet<string> Flag { get;set;}

    public string FlagsString
    {
        get
        {
            return Join(Flags, ",");
        }
        set
        {
            Flags = new HashSet<string>(Split(value, ","));
        }

The code uses the Flags fields, but when serialized the FlagsString property is used, converting Flags to/from a user friendly string.